File: 03.programmers/2021_kakao_blind_recuritment/01.C++/02.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <map>
#include <iostream>

using namespace std;

map<string, int> ordered;
bool check[20];

bool cmp(pair<string, int> a, pair<string, int> b)
{
    return a.second > b.second;
}

void power_set(string str, int n, int depth, int len)
{
    if(n == depth)
    {
        int i;
        string tmp = "";
        for(i = 0; i < n; i++)
        {
            if(check[i] == 1)
            {
                tmp.push_back(str[i]);
            }
        }
        if (tmp.length() >= len)
        {
            if (!ordered[tmp])
            {
                ordered.insert(pair<string, int>(tmp, 0));
            }
            ordered[tmp] += 1;
        }
        return;
    }
    check[depth]=1;
    power_set(str, n,depth+1, len);
    check[depth]=0;
    power_set(str, n,depth+1, len);
}

vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> answer;
    vector<string> tmp;
    
    for (auto order : orders)
    {
        sort(order.begin(), order.end());
        tmp.push_back(order);
    }
    
    for (int i = 0; i < orders.size(); i++)
        power_set(tmp[i], (int)tmp[i].length(), 0, 2);
    
    ordered.erase("");
    
    map<string, int>::iterator iter;
    
    for (int i = 0; i < course.size(); i++)
    {
        vector<pair<string, int>> vecs;
        vector<pair<string, int>> max;
        for (iter = ordered.begin(); iter != ordered.end(); iter++)
        {
            if (course[i] == iter->first.length() && iter->second > 1)
            {
                vecs.push_back(make_pair(iter->first, iter->second));
            }
        }
        if (!vecs.empty())
        {
            sort(vecs.begin(), vecs.end(), cmp);
            max.push_back(vecs[0]);
            for (auto vec : vecs)
            {
                cout << vec.first << " : " << vec.second << endl;
                if (max[0].second == vec.second)
                    answer.push_back(vec.first);
            }
        }
    }
    
    sort(answer.begin(), answer.end());
    
    return answer;
}

File: 2023/Programmers/lv01/javascript/바탕화면정리/02.js
function solution(wallpaper) {
  var answer = [];
  // 행 길이
  const n = wallpaper.length;
  // 열 길이
  const m = wallpaper[0].length;

  // 시작점
  let sy = n,
    sx = m;
  // 끝점
  let ey = 0,
    ex = 0;

  // 바탕화면 탐색
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      // 시작점을 찾기위한 정방향 탐색
      if (wallpaper[i][j] === "#") {
        sy = Math.min(sy, i);
        sx = Math.min(sx, j);
        ey = Math.max(ey, i + 1);
        ex = Math.max(ex, j + 1);
      }
    }
  }
  answer = [sy, sx, ey, ex];

  return answer;
}
